### PR TITLE
tests: Add fuzzing harness for CScript and CScriptNum operations

### PR DESCRIPTION
**Pull request by practicalswift**
Add fuzzing harness for `CScript` and `CScriptNum` operations.

Test this PR using:

```
$ make distclean
$ ./autogen.sh
$ CC=clang CXX=clang++ ./configure --enable-fuzz \
      --with-sanitizers=address,fuzzer,undefined
$ make
$ src/test/fuzz/script_ops
…
$ src/test/fuzz/scriptnum_ops
…
```

### CODE REVIEW

**Comment by MarcoFalke** [comment]
ACK 0730ac98c61f895893d40cb0fdff4cd7339a11b0

**Comment by practicalswift** [comment]
@MarcoFalke Added `CScriptNum` fuzzer too. Moved common functions to `fuzz.h`. Please re-review :)

File: src/test/fuzz/scriptnum_ops.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+{
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+    CScriptNum script_num = ConsumeScriptNum(fuzzed_data_provider);
+    while (fuzzed_data_provider.remaining_bytes() > 0) {
+        switch (fuzzed_data_provider.ConsumeIntegralInRange(0, 26)) {
+        case 0:
+            (void)(script_num == fuzzed_data_provider.ConsumeIntegral<int64_t>());
+            break;
+        case 1:
+            (void)(script_num != fuzzed_data_provider.ConsumeIntegral<int64_t>());
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Can combine case 0 and 1?

```cpp
const auto int = fuzzed_data_provider.ConsumeIntegral<int64_t>()
(script_num == int) != (script_num != int)
```

File: src/test/fuzz/scriptnum_ops.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            break;
+        case 2:
+            (void)(script_num <= fuzzed_data_provider.ConsumeIntegral<int64_t>());
+            break;
+        case 3:
+            (void)(script_num < fuzzed_data_provider.ConsumeIntegral<int64_t>());
+            break;
+        case 4:
+            (void)(script_num >= fuzzed_data_provider.ConsumeIntegral<int64_t>());
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Same here:

```cpp
(s_n < int) != (s_n >= int)

File: src/test/fuzz/scriptnum_ops.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            break;
+        case 5:
+            (void)(script_num > fuzzed_data_provider.ConsumeIntegral<int64_t>());
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Same here:

```cpp
(s_n > int ) != (s_n <= int)

File: src/test/fuzz/scriptnum_ops.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            (void)(script_num <= ConsumeScriptNum(fuzzed_data_provider));
+            break;
+        case 9:
+            (void)(script_num < ConsumeScriptNum(fuzzed_data_provider));
+            break;
+        case 10:
+            (void)(script_num >= ConsumeScriptNum(fuzzed_data_provider));
+            break;
+        case 11:
+            (void)(script_num > ConsumeScriptNum(fuzzed_data_provider));
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Same for all of these

File: src/test/fuzz/scriptnum_ops.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            break;
+        case 12:
+            script_num = script_num + fuzzed_data_provider.ConsumeIntegral<int64_t>();
+            break;
+        case 13:
+            script_num = script_num - fuzzed_data_provider.ConsumeIntegral<int64_t>();
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Could combine these?

```cpp
(script_num - int ) + int == script_num
(script_num + int ) - int == script_num

File: src/test/fuzz/scriptnum_ops.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            break;
+        case 14:
+            script_num = script_num + ConsumeScriptNum(fuzzed_data_provider);
+            break;
+        case 15:
+            script_num = script_num - ConsumeScriptNum(fuzzed_data_provider);
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Same here and below?

**Review by maflcko — APPROVED** [review]
ACK

File: src/test/fuzz/scriptnum_ops.cpp (hunk 1 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: Fixed!

File: src/test/fuzz/scriptnum_ops.cpp (hunk 3 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: Fixed!

File: src/test/fuzz/scriptnum_ops.cpp (hunk 5 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: Fixed!

File: src/test/fuzz/scriptnum_ops.cpp (hunk 6 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: Fixed!

File: src/test/fuzz/scriptnum_ops.cpp (hunk 2 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: Fixed!

File: src/test/fuzz/scriptnum_ops.cpp (hunk 4 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: Fixed!

**Comment by practicalswift** [comment]
@MarcoFalke Thanks for reviewing. All feedback addressed. Please re-review :)

**Comment by practicalswift** [comment]
Rebased again :)

File: src/test/fuzz/scriptnum_ops.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            break;
+        }
+        case 1: {
+            const int64_t i = fuzzed_data_provider.ConsumeIntegral<int64_t>();
+            assert((script_num <= i) != script_num > i);
+            break;
+        }
+        case 2: {
+            const int64_t i = fuzzed_data_provider.ConsumeIntegral<int64_t>();
+            assert((script_num >= i) != (script_num < i));
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: in commit 4532649328ad5056f9ae2bf99f3aba2212bdc785:

Any reason those are separate cases for the fuzzer to find? They can all be executed in the same case:

```cpp
             assert((script_num == i) != (script_num != i));
             assert((script_num <= i) != script_num > i);
             assert((script_num >= i) != (script_num < i));

File: src/test/fuzz/scriptnum_ops.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            break;
+        }
+        case 4: {
+            const CScriptNum sn = ConsumeScriptNum(fuzzed_data_provider);
+            assert((script_num <= sn) != (script_num > sn));
+            break;
+        }
+        case 5: {
+            const CScriptNum sn = ConsumeScriptNum(fuzzed_data_provider);
+            assert((script_num >= sn) != (script_num < sn));
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Same here

File: src/test/fuzz/scriptnum_ops.cpp [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                assert((script_num - i) + i == script_num);
+            }
+            break;
+        }
+        case 7: {
+            const CScriptNum sn = ConsumeScriptNum(fuzzed_data_provider);
+            // Avoid signed integer overflow:
+            // script/script.h:264:93: runtime error: signed integer overflow: -9223126527765971126 + -9223372036854756825 cannot be represented in type 'long'
+            if (IsValidAddition(script_num, sn)) {
+                assert((script_num + sn) - sn == script_num);
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Same here

File: src/test/fuzz/scriptnum_ops.cpp [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        case 17:
+            script_num &= fuzzed_data_provider.ConsumeIntegral<int64_t>();
+            break;
+        case 18: {
+            // Avoid negation failure:
+            // script/script.h:332:35: runtime error: negation of -9223372036854775808 cannot be represented in type 'int64_t' (aka 'long'); cast to an unsigned type to negate this value to itself
+            if (script_num == CScriptNum{std::numeric_limits<int64_t>::min()}) {
+                break;
+            }
+            (void)script_num.getvch();
```
Comment (MarcoFalke) [inline src/test/fuzz/scriptnum_ops.cpp]: Also in commit 4532649:

No input is consumed from the fuzzer, so this doesn't need to be special cased and can be executed unconditionally?

**Review by maflcko — APPROVED** [review]
ACK ad040fc2a503a0a5c7097dfe8aa6d341901436d9 👦

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK ad040fc2a503a0a5c7097dfe8aa6d341901436d9 👦
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgKHwv+OE49ZDbtGMchMWXzOuFyxStj7biNTlNtYQH2AIBvgxPl2zZbczejRMMA
xbUrKDh3SiE+mzPNyzOZ69XejxNGCTX1Kfel22F/XIwbKDw2kFv0v1F39pYyp4ZV
j7/CLVpqZaWdgqcdFvE5tYCBcflm/6bU95TabNkv3yjsZ0XEy7WGaGJTRsgeMoQd
wHIOc6kZerms8G04U/84PtHJBO4x0pU4zk2PRiEv9IcYYkAUFh1NDgIoJ+8MH9bt
uuSjVmOzKmbuc5mrq8VFbvgeHJlIBWtjNP7Ux2OkvUCxGeUDoMOluO4MlME/tZVQ
VJAQ6ZdOHSp032e5aOs9A9rwU+S6SlGxuoBlrjvK5vZm/nTwM8QXTU+hgza8hcdj
7IWPhHIFiEExdYRd31q/aZ7kU4fpp9GaD4jXq+5hiAFso2jkWTxAIH3sdIn0ilKp
+XxdLNNDSevd2KlCZK/DiwtTvU/Hp6WDPG9JW1YG0DS6jV1VeDmoacoN5M1hMYQz
8kqbBKQP
=dRzS
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `009db7b3b5d4284568143d3ed9fcbb2849ef578b5853b7f040cd0dd8c4f64f5f  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108009db7b3b5d4284568143d3ed9fcbb2849ef578b5853b7f040cd0dd8c4f64f5ff010c36e58d52ce3c1afd94fe280bb161dd508fff0103f889cb5929f47fa897cd8220128aa0608f1045e66a37ef0081583e78e12e255320083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01077450a81f675adda98b8d4bba104017708f0206e35970c55c0c97562ab18da900546149ac7f86dd8a7ce6bb6d3ff5c5c09ebf408f12092c7fb1d5b9231866b391a8abc8e0dfc40cd58a8d35b933b0e673aa28384e73508f1045e66a37ef008feada6aa153c0d790083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d08ef9479bde781353a5eb0e0fd8e04808f02095f8c63d354e74f0dd5bae0a7d9006b20daf0666a4446502ccf6b803d395835208f1045e66a37df0083931e9bd9b06dd750083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fc5cfcfd5ebdb230277f2e2b52ab764f08f0204734a9168a13e464b52894b47b277f401d5728d986513c726f661c33a3865b2808f1045e66a37ef008bb40002122ee85d60083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

File: src/test/fuzz/scriptnum_ops.cpp (hunk 7 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: Oh, of course. Now fixed. Thanks!

File: src/test/fuzz/scriptnum_ops.cpp (hunk 10 above)
Comment (practicalswift) [reply src/test/fuzz/scriptnum_ops.cpp]: True! Fixed.

**Comment by MarcoFalke** [comment]
ACK e37f53648e3acc6aea75adafec4de2bdbd8cb293 🦂

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e37f53648e3acc6aea75adafec4de2bdbd8cb293 🦂
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiTxwv/fFqtr5Cy14W/HZr4I58Tpz3S86mobLUDaulSWhFpMz8b38ilBC5ltwIH
rOznRDmgl46dP3YPMU07vCM2LS39DZyQH8uLTbECZw1aP5/x+XHb7me4aN/Ws1hB
tvD5zKvFcFnw+wbmkStYqDMaNKH2vFOSJKhWckcRouaB+fwZLNMzps1CH7tqixVg
Yo6i/PfETRSK9FveNFzbhSvhpGozLuFkqr0p5XyVWG42DxyOfe61VJP+e5RtlUFb
QUPiEZnm2UFgy3DFhPGw0y2Gz81kDvW5H3RD+nKxg2FmXeBuQlNpos/naAqq1pQ1
WsNtbwzJw5mrvxWvtALFFzvkBOv1W7kjes987ojGLQscMCtb1N/ZTb81sUSRzkW/
L5AkNgnyP8d6WFvzuztTgLt7ZxM/0hTJtIO3x1qC/VySZKR2Enhh/DrOACKQ8u0s
x3uTHh0e4qcOMz/6gKlMO2Boq48EUmUTNOSNY/95XY+0A5J7c6v6NuFIvhHq8Jfj
1N+3Xt0G
=AMUq
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `de3f5a712c02cb66b6ad6e76c0ec5a38c880d09962e085e811ec889daf153c65  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108de3f5a712c02cb66b6ad6e76c0ec5a38c880d09962e085e811ec889daf153c65f010d9d202833a4a1814d76dd3f03fb99db808fff01045402923a1e9b81cfda7912eac985c2c08f1045e6702f1f0088618730ab64c0d8b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0107d7f3303c7fa7f39a5488f64f6018d7f08f1045e6702f1f008bca925e1859d2e260083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010ae6b3fd474696d658726103f6683455908f1045e6702f1f0088fb1663f816c98ba0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ec523c2ef22bd5d5cdc61a7bb80bdc0908f1045e6702f1f0085515caa24d82d8550083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Pull request merged**